[Tizen] Update SDK version to 9.0, install.sh script improvements

enkiusz opened the pull request:
#### Summary

Bump the Tizen SDK version to 9.0. Some fixes and improvements were also applied to the Tizen-specific SDK "installer" adding missing dependencies, localising common code and separating the package download and install steps.

#### Testing

`python3 scripts/build/build_examples.py --target tizen-arm-tests-no-ble-no-thread build` has completed successfully. The example apps are also working on a Raspberry Pi 4 running Tizen 9.0 and have been verified to commission properly using the matter-qa framework.

#### Readability checklist

The checklist below will help the reviewer finish PR review in time and keep the
code readable:

-   [X] PR title is
        [descriptive](https://project-chip.github.io/connectedhomeip-doc/pull_request_guidelines.html#title-formatting)
-   [X] Apply the
        [_“When in Rome…”_](https://project-chip.github.io/connectedhomeip-doc/style/CODING_STYLE_GUIDE.html)
        rule (coding style)
-   [X] PR size is short
-   [X] Try to avoid "squashing" and "force-update" in commit history
-   [X] CI time didn't increase

enkiusz committed:
[Linux/Tizen] Update SDK version to 9.0, install.sh script improvements

Signed-off-by: Maciej Grela <m.grela@samsung.com>

restyled-commits committed:
Restyled by shellharden

enkiusz committed:
[Linux/Tizen] Improve help on missing dependencies

Signed-off-by: Maciej Grela <m.grela@samsung.com>

enkiusz committed:
[Linux/Tizen] Check for $TIZEN_SDK_ROOT being empty before trying to perform the unzip data/ symlink trick

Signed-off-by: Maciej Grela <m.grela@samsung.com>

enkiusz committed:
[Linux/Tizen] Improve tizen_qemu.py

- warn when pigweed's qemu-system-arm is the only one we have
- apply best practices for `logging` use
- pylint fixes

Signed-off-by: Maciej Grela <m.grela@samsung.com>

restyled-commits committed:
Restyled by autopep8

enkiusz committed:
[Linux/Tizen] Harmonize help string syntax

Signed-off-by: Maciej Grela <m.grela@samsung.com>

restyled-commits committed:
Restyled by autopep8

arkq commented on integrations/docker/images/stage-2/chip-build-tizen/tizen-sdk-installer/install.sh:
```diff
@@ -48,6 +49,9 @@ function show_help() {
     echo "  --tizen-sdk-data-path      Set directory for Tizen SDK runtime data. Default is $TIZEN_SDK_DATA_PATH"
+    echo "  --tizen-pkgs-dir           Set the directory to store downloaded Tizen SDK packages, default is /tmp/XXXXXX"
+    echo "  --fetch-sdk-pkgs           Fetch RPMs and ZIPs from https://download.tizen.org/ into pkgs path"
+    echo "  --install-sdk-pkgs         Install fetched packages into Tizen SDK directory (default $TIZEN_SDK_ROOT)"
```
How about fetching and always storing packages in the cache dir (maybe a subdir in `~/.cache` or some other common persistent "cache" directory, just like `~/.cache/pip`). Then, when package is already in the cache skip the fetch stage (for a single package). The install stage should always happen (the name of this script is `install.sh`). Optionally you might add an option to always-fetch even though a package is in the cache already (alternatively one can simply remove the cache dir to achieve the same result).

arkq commented on third_party/tizen/tizen_qemu.py:
```diff
@@ -153,7 +159,7 @@ def whereis(binary_name):
 # Run QEMU.
 with open(args.output, "wb") as output:
-    logging.info("run: %s", " ".join(map(shlex.quote, qemu_args)))
+    log.info("run: %s", " ".join(map(shlex.quote, qemu_args)))
```
I think that by default the info level is not shown. However, this "run: ..." log message is really helpful when debugging issues (so one can run qemu manually). Please make sure that it is visible.

enkiusz committed:
[Linux/Tizen] Improvements to Tizen SDK install.sh

- implify usage, the tool now behaves like python's pip leaving behind a cache directory and reusing it instead of downloading files every time. This behaviour can be controlled by the --purge-pkg-cache which starts an installation from a clean cache
- have more control over script verbosity with commandline flags and DEBUG environment variable
- messages passed to printing functions info(), error(), etc. do not have to be a single arg anymore
- use explicit parameters to pass output paths to wget instead of relying on `cd` - less magical context in the script
- handle unpacking of zips and RPMs where multiple versions of a single component are present in the cache - the most recent file is selected
- don't force unzip to automatically overwrite files, this was weird and was likely needed because one zip was unpacked twice
- generate rpm lists for arm and arm64 architecture from a common source of truth

Signed-off-by: Maciej Grela <m.grela@samsung.com>

restyled-commits committed:
Restyled by shellharden

enkiusz committed:
[Linux/Tizen] convert chip-build-tizen Dockerfile to new install.sh syntax

Signed-off-by: Maciej Grela <m.grela@samsung.com>

enkiusz committed:
[Linux/Tizen] Further install.sh fixes and improvements

- pass arrays properly to unpack functions
- secure against shellharden injecting empty argv items into unzip calls, unzip does not like that

Signed-off-by: Maciej Grela <m.grela@samsung.com>

restyled-commits committed:
Restyled by whitespace

restyled-commits committed:
Restyled by shellharden

enkiusz committed:
[Linux/Tizen] Make the script pass shellharden without issues

Signed-off-by: Maciej Grela <m.grela@samsung.com>

enkiusz committed:
install.sh: Fix typo

Signed-off-by: Maciej Grela <m.grela@samsung.com>

restyled-commits committed:
Restyled by shfmt

arkq commented on integrations/docker/images/stage-2/chip-build-tizen/tizen-sdk-installer/install.sh:
```diff
@@ -63,47 +76,52 @@ function show_help() {
+    wget -r "$verbose_flag" -nd --timestamping --no-parent -e robots=off --progress=dot:mega -P "$PKG_CACHE_PATH" "${PKGS[@]}" "$url"
 
     # Check if the files have been downloaded
     for PKG in "${@:2}"; do
```
Few lines above you've changed `"${@:2}"` to `"${@}"` bot not here.

Also, it might be work to mark `PKG` and `PKGS` as local variables (and make them lowercase).

arkq commented on integrations/docker/images/stage-2/chip-build-tizen/tizen-sdk-installer/install.sh:
```diff
@@ -127,290 +145,275 @@ function install_dependencies() {
+
+        info "Unpacking '$most_recent_filename' into '$destdir'"
+
+        # Shellharden puts quotes around ${VERBOSE:+-v} which causes an empty
```
The comment says that it's not possible to use `-v` as alternative value, but then unzip is called without `-v`...

enkiusz replied:
The comment might be a bit unclear in that case. The problem is that when using "${VERBOSE+-v}" we end up with an empty string if $VERBOSE is unset. When unzip gets a first argument which is an empty string it thinks it's a zip file tries to extract it and fails. See:

```
➜  ~ file instructions.zip 
instructions.zip: Zip archive data, at least v2.0 to extract, compression method=deflate
➜  ~ cat t.sh 
unset VERBOSE
[ "$1" == "verbose" ] && VERBOSE=true

unzip "${VERBOSE+-v}" instructions.zip

➜  ~ bash -x t.sh
+ unset VERBOSE
+ '[' '' == verbose ']'
+ unzip '' instructions.zip
unzip:  cannot find or open , .zip or .ZIP.
➜  ~ bash -x t.sh verbose
+ unset VERBOSE
+ '[' verbose == verbose ']'
+ VERBOSE=true
+ unzip -v instructions.zip
Archive:  instructions.zip
 Length   Method    Size  Cmpr    Date    Time   CRC-32   Name
--------  ------  ------- ---- ---------- ----- --------  ----
    1250  Defl:N      489  61% 2024-12-21 13:14 708e12ba  instructions.txt
--------          -------  ---                            -------
    1250              489  61%                            1 file
➜  ~ 
```

enkiusz commented on integrations/docker/images/stage-2/chip-build-tizen/tizen-sdk-installer/install.sh:
```diff
@@ -127,290 +145,275 @@ function install_dependencies() {
+# unzip_globs function below wrt. globs matching multiple RPMs.
+function unrpm_globs() {
+    local destdir="$1"
+    declare -n globs="$2"
```
`globs` is an array argument, it's cleanest to pass it with a reference, no need to expand it to a list when calling.

enkiusz committed:
Address review comments

- do not purge when building docker images
- declare all local vars with `local`
- convert all local var names to lowercase
- fix wrong use of ${@} in dowload()

Signed-off-by: Maciej Grela <m.grela@samsung.com>

restyled-commits committed:
Restyled by shellharden

arkq commented on integrations/docker/images/stage-2/chip-build-tizen/Dockerfile:
```diff
@@ -1,5 +1,5 @@
 ARG VERSION=1
-FROM ghcr.io/project-chip/chip-build:${VERSION}
+FROM ghcr.io/enkiusz/chip-build:${VERSION}
```
???

enkiusz committed:
Use proper base image in chip-build-tizen

Signed-off-by: Maciej Grela <m.grela@samsung.com>

enkiusz committed:
Merge remote-tracking branch 'upstream/master' into tizen-9.0

enkiusz committed:
Bump Docker image version

Signed-off-by: Maciej Grela <m.grela@samsung.com>
